flax/errors.py

```diff
@@ -298,4 +298,297 @@ def __call__(self, x):
+# module.py errors                              #
+#################################################
+
+
+class  ModuleNameExistsError(FlaxError):
+  """
+  This error is raised when trying to create a submodule with an exsiting name.
+  Witin a scope, submodules should have unique names, because each of them 
+  are stored as unique keys in the variable dict (for more explanation on 
+  variable dicts,  please see :mod:`flax.core.variables`.)::
+
+    class Foo(nn.Module):
+      @nn.compact
+      def __call__(self, x):
+        dense = nn.Dense(features=3, name='bar')
+        embed = nn.Embed(num_embeddings=2, features=5, name='bar')  # <-- ERROR!
+
+    Foo().init(random.PRNGKey(0), np.arange(5))
+
+  Similarly, a submodule name can collide with an existing parameter name, since
+  they are both stored in the same variable dict::
+
+    class Foo(nn.Module):
+      @nn.compact
+      def __call__(self, x):
+        bar = self.param('bar', nn.initializers.zeros, (1, ))
+        embed = nn.Embed(num_embeddings=2, features=5, name='bar')  # <-- ERROR!
+  """
+  def __init__(self, module_name, variable_name):
+    super().__init__(f'A module of name "{variable_name}" exists already.')
+
+
+class AssignSubModuleError(FlaxError):
+  """
+  You are only allowed to create submodules in two places:
+
+  1.  If your Module is noncompact: inside 
+      :meth:`Module.setup() <flax.linen.Module.apply>`.
+  2.  If your Module is compact: inside the method wrapped in 
+      :meth:`nn.compact() <flax.linen.compact>`.
+
+  For instance, the following code throws this error, because ``nn.Conv`` is
+  created in ``__call__`, which is not marked as compact.
+
+    class Foo(nn.Module):
+      def setup(self):
+        pass
+
+      def __call__(self, x):
+        conv = nn.Conv(features=3, kernel_size=3)
+
+  Foo().init(random.PRNGKey(0), jnp.zeros((1,)))
+
+  Note that this error is also thrown if you partially defined a Module inside
+  setup::
+
+    class Foo(nn.Module):
+    def setup(self):
+      self.conv = functools.partial(nn.Conv, features=3)
+
+    def __call__(self, x):
+      x = self.conv(kernel_size=4)(x)
+      return x
+
+    Foo().init(random.PRNGKey(0), jnp.zeros((1,)))
+
+  In this case, ``self.conv(...)`` inside ``__call__`` constructs a 
+  ``ConvBlock``, which will check where inside the parent Module ``Foo`` it was
+  created, and realize this was outside of setup.
+  """
+  def __init__(self, cls):
+    super().__init__(f'Submodule {cls} must be defined in `setup()` or in a '
+                     'method wrapped in `@compact`')
+
+
+class ModuleAttributeNotFoundError(FlaxError):
+  """
+  This error occurs when trying to access a module attribute that does not
+  exist. For instance, the following code will throw the error because attribute
+  ``unknown`` is not defined::
+
+    class Foo(nn.Module):
+      @nn.compact
+      def __call__(self, x):
+        y = self.unknown  # <--- ERROR
```

Comment:
(But maybe there's no way to catch this error otherwise?)

Reply:
I'm not sure what you mean, this is normal behavior right? We are simply accessing an object attribute that doesn't exist. Consider the following Python code:

```
class Test:
  def hoi(self):
    print(self.unknown)

Test().hoi()
```

That also throws an error: `AttributeError: 'Test' object has no attribute 'unknown'`

Or do you mean it is surprising that we raise a custom error?

Reply:
Yes that it is surprising that we raise a custom error here.

Reply:
I agree, but the main reason I added is so that we can explain this problem better: #1014

Reply:
I understand, and perhaps there's no good solution here. The problem is that it catches the error too broadly. Most likely someone just had a typo and it's unrelated to whether they used `apply` or not...

Reply:
Yes I agree, the current solution is a bit unnatural. It would be better if we can catch the error in #1014 in another way and not abuse AttributeError for it. I've reverted it.